cli/command/trust/key_generate.go

```diff
@@ -0,0 +1,131 @@
+package trust
+
+import (
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/trust"
+	"github.com/docker/notary"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/data"
+	tufutils "github.com/docker/notary/tuf/utils"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+type keyGenerateOptions struct {
+	name      string
+	directory string
+}
+
+func newKeyGenerateCommand(dockerCli command.Streams) *cobra.Command {
+	options := keyGenerateOptions{}
+	cmd := &cobra.Command{
+		Use:   "generate NAME [NAME...]",
+		Short: "Generate and load a signing key-pair",
+		Args:  cli.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.name = args[0]
+			return setupPassphraseAndGenerateKeys(dockerCli, options)
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&options.directory, "dir", "d", "", "Directory to generate key in, defaults to current directory")
+	return cmd
+}
+
+// key names can use alphanumeric + _ + - characters
+var validKeyName = regexp.MustCompile(`^[a-zA-Z0-9\_]+[a-zA-Z0-9\_\-]*$`).MatchString
+
+// validate that all of the key names are unique and are alphanumeric + _ + -
+// and that we do not already have public key files in the current dir on disk
+func validateKeyArgs(keyName string, cwdPath string) error {
+	if !validKeyName(keyName) {
+		return fmt.Errorf("key name \"%s\" must not contain special characters", keyName)
+	}
+
+	pubKeyFileName := keyName + ".pub"
+	if _, err := os.Stat(filepath.Join(cwdPath, pubKeyFileName)); err == nil {
+		return fmt.Errorf("public key file already exists: \"%s\"", pubKeyFileName)
+	}
+	return nil
+}
+
+func setupPassphraseAndGenerateKeys(streams command.Streams, opts keyGenerateOptions) error {
+	targetDir := opts.directory
+	// if the target dir is empty, default to CWD
```

Comment:
This comment is a bit redundant as it's just the code "in words", also leaning towards having the validation in this step, instead of further down in `validateAndGenerateKey()`, something like:

```patch
diff --git a/cli/command/trust/key_generate.go b/cli/command/trust/key_generate.go
index 913083be..2cb00401 100644
--- a/cli/command/trust/key_generate.go
+++ b/cli/command/trust/key_generate.go
@@ -59,25 +59,23 @@ func validateKeyArgs(keyName string, cwdPath string) error {
 }

 func setupPassphraseAndGenerateKeys(streams command.Streams, opts keyGenerateOptions) error {
-       targetDir := opts.directory
-       // if the target dir is empty, default to CWD
-       if targetDir == "" {
+       if opts.directory == "" {
                cwd, err := os.Getwd()
                if err != nil {
                        return err
                }
-               targetDir = cwd
+               opts.directory = cwd
        }
-       return validateAndGenerateKey(streams, opts.name, targetDir)
-}
-
-func validateAndGenerateKey(streams command.Streams, keyName string, workingDir string) error {
-       freshPassRetGetter := func() notary.PassRetriever { return trust.GetPassphraseRetriever(streams.In(), streams.Out()) }
-       if err := validateKeyArgs(keyName, workingDir); err != nil {
+       if err := validateKeyArgs(opts.name, opts.directory); err != nil {
                return err
        }
+       return generateKey(streams, opts.name, opts.directory)
+}
+
+func generateKey(streams command.Streams, keyName string, workingDir string) error {
        fmt.Fprintf(streams.Out(), "\nGenerating key for %s...\n", keyName)
        // Automatically load the private key to local storage for use
+       freshPassRetGetter := func() notary.PassRetriever { return trust.GetPassphraseRetriever(streams.In(), streams.Out()) }
        privKeyFileStore, err := trustmanager.NewKeyFileStore(trust.GetTrustDirectory(), freshPassRetGetter())
        if err != nil {
                return err
```